### PR TITLE
Fix __future__ import behaviour in Workbench scripts

### LINKED ISSUE #27019: Workbench __future__ imports still uses old functionality
James Lord has been testing Workbench Mantid 4.1 functionality.

He has tested from __future__ import (print_function, division, unicode_literals) 

and as these 3 do not work correctly, it seems that 'absolute_import' is likely not to work either.

See below for details:

1.	The “from future import print_function” for Python 3 preparations isn’t working as advertised. If I use print (“the answer is”,x) I get a tuple in the messages window such as (‘the answer is’,42.0). It’s as if the future statement is used for syntax checking, rejecting bare print statements, but then turns itself off.

***YES Works in Plot, **not** in Workbench***

2.	To follow up, this works under standard Python 2.x, even using …/MantidInstall/bin/python.exe, but NOT in Workbench (gives a syntax error):
from __future__ import print_function
myprint=print
myprint ( "the answer is", 42 )

***YES Works in Plot, **not** in Workbench***

3.	And it appears the other __ future__ statements are broken too.
“division” doesn’t make 1/3 return 0.33333 (it’s still integer 0 as in plain Python 2.7).

***YES Works in Plot, **not** in Workbench***

4.	unicode_literals is broken too. String literals are not Unicode unless I add the “u” prefix. If I attempt to save a script with a non-ascii character in a string literal, such as £ (with or without a source code encoding indicator as in PEP263) I get an error traceback indicating it was trying to use ascii encoding and the error reporting dialog. And if I put in the encoding indicator and then try to run the (unsaved) script I get a syntax error.

***YES save in Workbench throws up an Error report! Mostly only seems a problem when saving!***

### PR DESCRIPTION
**Pull request by ghost**
**Description of work.**
The \_\_future\_\_ imports were not doing their job in scripts run in Workbench because the relevant compiler flags were not passed to the `compile` function when the scripts were executed.

This PR parses the given code for \_\_future\_\_  imports and gives the relevant flags to the `compile` function before its output is executed.

**To test:**
The following should run without error, and as described, in Workbench:
```python
# The following line helps with future compatibility with Python 3
# print must now be used as a function, e.g print('Hello','World')
from __future__ import (absolute_import, division, print_function, unicode_literals)

print("Fixed", "if", "no", "brackets", "printed")
print("Fixed if following is not zero", 1/3)
print("This (non-ASCII) pound sign should print: £")
print()
```

**Note**: you will not be able to save the above script, there is a separate issue for this here #27072 

Fixes #27019 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/widgets/codeeditor/execution.py [hunk 1]
```diff
@@ -90,15 +126,16 @@ def execute(self, code_str, filename=None):
             self.globals_ns[FILE_ATTR] = filename
         else:
             filename = EMPTY_FILENAME_ID
-        compile(code_str, filename, mode=COMPILE_MODE, dont_inherit=True)
+        flags = get_future_import_compiler_flags(code_str)
+        compile(code_str, filename, mode=COMPILE_MODE, dont_inherit=True, flags=flags)
```
Comment (hsautessella) [inline qt/python/mantidqt/widgets/codeeditor/execution.py]: Looking at this again, is this line actually doing anything? As far as I can tell, it isn't.

Comment (hsautessella) [reply qt/python/mantidqt/widgets/codeeditor/execution.py]: Ahh, it's checking for syntax errors in the whole script. When we don't have this line, the code will run up to the line with the syntax error. I'll add a comment so no one is tempted to remove it

File: qt/python/mantidqt/widgets/codeeditor/execution.py [hunk 2]
```diff
@@ -23,6 +25,40 @@
...
 
+def _get_imported_from_future(code_str):
+    """
+    Parse the given code and return a list of names that are imported
+    from __future__.
+    :param code_str: The code to parse
+    :return list: List of names that are imported from __future__
+    """
+    future_imports = []
+    for node in ast.walk(ast.parse(code_str)):
```
Comment (martyngigg) [inline qt/python/mantidqt/widgets/codeeditor/execution.py]: Nice use of the `ast` module :+1:

File: qt/python/mantidqt/widgets/codeeditor/execution.py (hunk 1 above)
Comment (martyngigg) [reply qt/python/mantidqt/widgets/codeeditor/execution.py]: :+1:

**Review by martyngigg — DISMISSED** [review]
Looks good. I tried a variety of combinations of `__future__` imports and they are behave as expected when present and not present.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**